File: 0229 majority-element-ii/solution.ts
function majorityElement(nums: number[]): number[] {
    let c1=0,c2=0;
    let cd1=0,cd2=0;
    for(const a of nums) {
        if(cd1==a) c1++;
        else if(cd2==a) c2++;
        else if(c1==0) {
            c1=1;
            cd1=a;
        } else if(c2==0) {
            c2=1;
            cd2=a;
        } else {
            c1--;
            c2--;
        }
    }
    c1=0;
    c2=0;
    for(const a of nums) {
        if(a==cd1) c1++;
        if(a==cd2) c2++;
    }
    let ans = [];
    if(c1 > nums.length/3) ans.push(cd1);
    if(c2 > nums.length/3 && cd2!=cd1) ans.push(cd2);
    return ans;
};

File: 0229 majority-element-ii/solution.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int c1 = 0, cd1 = 0, c2 = 0, cd2 = 0;
        for(int i:nums) {
            if(cd1==i) c1++;
            else if(cd2==i) c2++;
            else if(c1==0) {
                cd1=i;
                c1++;
            } else if(c2==0) {
                cd2=i;
                c2++;
            } else {
                c1--;
                c2--;
            }
        }
        vector<int> ans;
        c1 = 0, c2 = 0;
        for(int i:nums) {
            if(cd1 == i) c1++;
            else if(cd2 == i) c2++;
        }
        if(c1 > nums.size() /3) ans.push_back(cd1);
        if(c2 > nums.size() /3) ans.push_back(cd2);
        return ans;
    }
};
